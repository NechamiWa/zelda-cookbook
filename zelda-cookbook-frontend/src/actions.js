// Recipes =================

export const populateRecipes = recipes => ({
  type: 'POPULATE_RECIPES',
  recipes
});

// Ingredients =============

export const populateIngredients = ingredients => ({
  type: 'POPULATE_INGREDIENTS',
  ingredients
});

// Filter ==================

export const addIngredientFilter = ingredient => ({
  type: 'ADD_INGREDIENT_FILTER',
  ingredient
});

export const removeIngredientFilter = id => ({
  type: 'REMOVE_INGREDIENT_FILTER',
  id
});

export const updateNameFilter = name => ({
  type: 'UPDATE_NAME_FILTER',
  name
});

export const addTypeFilter = recipeType => ({
  type: 'ADD_TYPE_FILTER',
  recipeType
});

export const removeTypeFilter = () => ({
  type: 'REMOVE_TYPE_FILTER'
});