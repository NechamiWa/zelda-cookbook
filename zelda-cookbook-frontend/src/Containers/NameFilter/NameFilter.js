import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { updateNameFilter } from '../../actions';

let timeout;

export class NameFilter extends Component {
  constructor(props) {
    super(props);
    this.state = {
      recipeSearch: ''
    };
  }

  handleChange = event => {
    const recipeSearch = event.target.value;
    this.setState({ recipeSearch });
    clearTimeout(timeout);
    timeout = setTimeout(() => {
      this.props.updateNameFilter(recipeSearch);
    }, 500);
  }

  resetSearch = () => {
    this.props.updateNameFilter('');
    this.setState({ recipeSearch: ''});
  }

  render() {
    const { recipeSearch } = this.state;
    return (
      <div className='NameFilter'>
        <input
          type="text"
          placeholder="Recipe Name"
          value={this.state.recipeSearch}
          className='search-bar search-bar__recipe'
          onChange={this.handleChange}
        />
        <button
          onClick={ this.resetSearch }
          className={`button__clear-filter
            ${ !recipeSearch ? 'button__clear-filter--hidden' : ''}`}
        >
          X
        </button>
      </div>
    );
  }
}

const { func } = PropTypes;

NameFilter.propTypes = {
  updateNameFilter: func
};

export const MDTP = dispatch => ({
  updateNameFilter: name => dispatch(updateNameFilter(name))
});

export default connect(null, MDTP)(NameFilter);