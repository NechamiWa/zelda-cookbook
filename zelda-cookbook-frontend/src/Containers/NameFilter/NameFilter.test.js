/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { NameFilter, MDTP } from './NameFilter';
import { updateNameFilter } from '../../actions';

jest.useFakeTimers();

describe('NameFilter', () => {
  let wrapper;
  let mockUpdateNameFilter = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<NameFilter 
      updateNameFilter = { mockUpdateNameFilter }
    />)
  })

  it('should have a default state', () => {
    const expected = {
      recipeSearch: '',
    };

    expect(wrapper.state()).toEqual(expected);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleChange', () => {
    it('should set state with a new recipe name', () => {
      const event = { target: { 
        value: 'asdf'
      }};

      wrapper.instance().handleChange(event);
      expect(wrapper.state().recipeSearch).toEqual('asdf');
    });

    it('should wait until the user is finished typing, then call updateNameFilter with the value in recipeSearch', () => {
      const event = { target: { 
        name: 'recipeSearch',
        value: 'Pie'
      }};
      wrapper.instance().handleChange(event)
      jest.runAllTimers()
      expect(mockUpdateNameFilter).toHaveBeenCalledWith('Pie')
    });
  });

  describe('resetSearch', () => {
    it('should call updateNameFilter with an empty string', () => {
      wrapper.instance().resetSearch();
      expect(mockUpdateNameFilter).toHaveBeenCalledWith('');
    });

    it('should set this.state.recipeSearch to an empty string', () => {
      wrapper.setState({ recipeSearch: 'asdf' })
      expect(wrapper.state().recipeSearch).toEqual('asdf')
      wrapper.instance().resetSearch();
      expect(wrapper.state().recipeSearch).toEqual('')
    });
  });
});

describe('MDTP', () => {
  it('should call dispatch when updateNameFilter is called', () => {
    const mockDispatch = jest.fn();
    const mockName = 'Pie';
    const result = MDTP(mockDispatch);

    result.updateNameFilter(mockName);
    expect(mockDispatch).toHaveBeenCalledWith(updateNameFilter(mockName));
  });
});