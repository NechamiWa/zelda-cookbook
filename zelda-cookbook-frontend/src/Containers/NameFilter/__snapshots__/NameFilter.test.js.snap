// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameFilter should match the snapshot 1`] = `
<div
  className="NameFilter"
>
  <input
    className="search-bar search-bar__recipe"
    onChange={[Function]}
    placeholder="Recipe Name"
    type="text"
    value=""
  />
  <button
    className="button__clear-filter
              button__clear-filter--hidden"
    onClick={[Function]}
  >
    X
  </button>
</div>
`;
